eyecite/count.py

```diff
@@ -0,0 +1,162 @@
+                    candidates.append(o)
+
+        refined_candidates = list(set(refined_candidates))
+        if len(refined_candidates) == 1:
+            return refined_candidates[0]
+        else:
+            return None
+
+
+def _match_fullcase_citation(citation: FullCaseCitation) -> Opinion:
+    """
+    Use fullcase citations as direct opinion matches.
+    """
+    return Opinion(citation)
+
+
+def count_citations(
+    citations: List[CitationBase],
+    match_supra_citation: Callable[
+        [SupraCitation, List[Any]], Optional[BaseOpinion]
+    ] = _match_supra_citation,
+    match_shortcase_citation: Callable[
+        [ShortCaseCitation, List[Any]], Optional[BaseOpinion]
+    ] = _match_shortcase_citation,
+    match_fullcase_citation: Callable[
+        [FullCaseCitation], BaseOpinion
+    ] = _match_fullcase_citation,
+    successful_match_callback: Optional[
+        Callable[[BaseOpinion, CitationBase], Any]
+    ] = None,
+) -> Dict[BaseOpinion, int]:
+    """Counts the number of unique opinions referenced by a list of citations
+    by matching each type of Citation object (FullCaseCitation,
+    ShortCaseCitation, SupraCitation, and IdCitation) to a thin opinion
+    object. Assumes that the list of citations is ordered in the order that
+    they were extracted from the text (i.e., assumes that supra citations and
+    id citations can only refer to previous references). Returns a dict in the
+    following format:
+        keys = cited opinion
+        values = number of times that opinion is cited
+
+    The individual matching steps can be supplanted with more complex logic
+    by passing custom functions (e.g., if you have a thicker opinion
+    abstraction that you want to use for matching); the default approach is to
+    use simple heuristics to narrow down the set of possible resolutions. If a
+    citation cannot be definitively resolved to an opinion, it is dropped and
+    not counted.
+    """
+    opinions: List[BaseOpinion] = []  # List of opinion-matches
+    was_matched: bool = (
+        False  # Whether the previous citation match was successful
+    )
+
+    # Iterate over each citation in order and attempt to match it to an opinion
+    for citation in citations:
+        matched_opinion: Optional[BaseOpinion] = None
+
+        # If the citation is to a non-opinion document, ignore
+        if isinstance(citation, NonopinionCitation):
+            pass
+
+        # If the citation is a full case citation, it can be matched to an
+        # opinion directly
+        elif isinstance(citation, FullCaseCitation):
+            matched_opinion = match_fullcase_citation(citation)
+
+        # If the citation is an id citation, just resolve it to the opinion
+        # that was matched immediately prior (so long as the previous match
+        # was successful)
+        elif isinstance(citation, IdCitation):
+            if was_matched:
+                matched_opinion = opinions[-1]
```

Comment:
Seems like this should get abstracted out too?

Reply:
Touché